OvmfPkg/VirtioKeyboardDxe/VirtioKeyboard.c

```diff
@@ -0,0 +1,1560 @@
+    [KEY_BACKSLASH]     = '|',
+    [KEY_Z]             = 'Z', 'X',  'C', 'V', 'B', 'N', 'M',
+    [KEY_COMMA]         = '<', '>',  '?',
+    [KEY_SPACE]         = ' ',
+    [MAX_KEYBOARD_CODE] = 0x00
+  };
+
+  // Check if key code is not out of the keyboard mapping boundaries
+  if (Code >= MAX_KEYBOARD_CODE) {
+    DEBUG ((DEBUG_INFO, "%a: Key code out of range \n", __func__));
+    Key->ScanCode    = SCAN_NULL;
+    Key->UnicodeChar = CHAR_NULL;
+    return;
+  }
+
+  // Handle F1 - F10 keys
+  if ((Code >= KEY_F1) && (Code <= KEY_F10)) {
+    Key->ScanCode    = SCAN_F1 + (Code - KEY_F1);
+    Key->UnicodeChar = CHAR_NULL;
+    return;
+  }
+
+  switch (Code) {
+    case KEY_PAGEUP:
+      Key->ScanCode    = SCAN_PAGE_UP;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_PAGEDOWN:
+      Key->ScanCode    = SCAN_PAGE_DOWN;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_HOME:
+      Key->ScanCode    = SCAN_HOME;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_END:
+      Key->ScanCode    = SCAN_END;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_DELETE:
+      Key->ScanCode    = SCAN_DELETE;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_INSERT:
+      Key->ScanCode    = SCAN_INSERT;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_UP:
+      Key->ScanCode    = SCAN_UP;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_LEFT:
+      Key->ScanCode    = SCAN_LEFT;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_RIGHT:
+      Key->ScanCode    = SCAN_RIGHT;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_DOWN:
+      Key->ScanCode    = SCAN_DOWN;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_BACKSPACE:
+      Key->ScanCode    = SCAN_NULL;
+      Key->UnicodeChar = CHAR_BACKSPACE;
+      break;
+
+    case KEY_TAB:
+      Key->ScanCode    = SCAN_NULL;
+      Key->UnicodeChar = CHAR_TAB;
+      break;
+
+    case KEY_ENTER:
+      Key->ScanCode = SCAN_NULL;
+      // Key->UnicodeChar = CHAR_LINEFEED;
+      Key->UnicodeChar = CHAR_CARRIAGE_RETURN;
+      break;
+
+    case KEY_ESC:
+      Key->ScanCode    = SCAN_ESC;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    default:
+      if (Dev->ShiftActive) {
+        Key->ScanCode    = MapShift[Code];
+        Key->UnicodeChar = MapShift[Code];
+      } else {
+        Key->ScanCode    = Map[Code];
+        Key->UnicodeChar = Map[Code];
+      }
+
+      if (Dev->CtrlActive) {
+        // Convert Ctrl+[a-z] and Ctrl+[A-Z] into [1-26] ASCII table entries
+        // Key->ScanCode    = SCAN_NULL;
+        if ((Key->UnicodeChar >= L'a') && (Key->UnicodeChar <= L'z')) {
+          Key->UnicodeChar = (CHAR16)(Key->UnicodeChar - L'a' + 1);
+        } else if ((Key->UnicodeChar >= L'A') && (Key->UnicodeChar <= L'Z')) {
+          Key->UnicodeChar = (CHAR16)(Key->UnicodeChar - L'A' + 1);

```

Comment:
You can just mask out the high bits, i.e. `UnicodeChar &= 0x1f`, and I think you can even do that unconditionally (not restricted to letters).  `Ctrl + ]` is a valid key combination for example.

Reply:
Done